src/infer_type.rs

```diff
@@ -0,0 +1,131 @@
+use crate::types::{Dimension, GeometryType};
+
+const POINT: &str = "POINT";
+const LINESTRING: &str = "LINESTRING";
+const POLYGON: &str = "POLYGON";
+const MULTIPOINT: &str = "MULTIPOINT";
+const MULTILINESTRING: &str = "MULTILINESTRING";
+const MULTIPOLYGON: &str = "MULTIPOLYGON";
+const GEOMETRYCOLLECTION: &str = "GEOMETRYCOLLECTION";
+
+/// Infer the geometry type and dimension from an input WKT string slice.
+///
+/// An `EMPTY` WKT object will return `None` in place of the dimension.
+///
+/// ```
+/// use wkt::infer_type;
+/// use wkt::types::{Dimension, GeometryType};
+///
+/// assert_eq!(
+///     infer_type("POINT (10 20.1)").unwrap(),
+///     (GeometryType::Point, Some(Dimension::XY))
+/// );
+///
+/// assert_eq!(
+///     infer_type("POINT EMPTY").unwrap(),
+///     (GeometryType::Point, None)
+/// );
+/// ```
+pub fn infer_type(input: &str) -> Result<(GeometryType, Option<Dimension>), String> {
+    if let Some((prefix, _suffix)) = input.split_once("(") {
+        let prefix = prefix.to_uppercase();
+
+        let (geom_type, dim_str) = if let Some(dim_str) = prefix.strip_prefix(POINT) {
+            (GeometryType::Point, dim_str)
+        } else if let Some(dim_str) = prefix.strip_prefix(LINESTRING) {
+            (GeometryType::LineString, dim_str)
+        } else if let Some(dim_str) = prefix.strip_prefix(POLYGON) {
+            (GeometryType::Polygon, dim_str)
+        } else if let Some(dim_str) = prefix.strip_prefix(MULTIPOINT) {
+            (GeometryType::MultiPoint, dim_str)
+        } else if let Some(dim_str) = prefix.strip_prefix(MULTILINESTRING) {
+            (GeometryType::MultiLineString, dim_str)
+        } else if let Some(dim_str) = prefix.strip_prefix(MULTIPOLYGON) {
+            (GeometryType::MultiPolygon, dim_str)
+        } else if let Some(dim_str) = prefix.strip_prefix(GEOMETRYCOLLECTION) {
+            (GeometryType::GeometryCollection, dim_str)
+        } else {
+            return Err(format!("Unsupported WKT prefix {}", prefix));
+        };
+
+        let dim = if dim_str.contains("ZM") {
+            Dimension::XYZM
+        } else if dim_str.contains("Z") {
+            Dimension::XYZ
+        } else if dim_str.contains("M") {
+            Dimension::XYM
+        } else {
+            Dimension::XY
+        };
+
+        Ok((geom_type, Some(dim)))
+    } else {
+        let input = input.to_uppercase();
+        if !input.contains("EMPTY") {
+            return Err("Invalid WKT; no '(' character and not EMPTY".to_string());
+        }
+
+        if input.starts_with(POINT) {
```

Comment:
input = " POINT EMTPY" 

(leading space?)

Reply:
Yeah... do you think that happens often? Is that valid WKT?

Reply:
Would this be better as a `contains`?

Reply:
> Is that valid WKT?

Your guess is as good as mine.

But apparently we support parsing it: 

<img width="697" alt="Screenshot 2024-10-07 at 13 06 20" src="https://github.com/user-attachments/assets/8c9f99c1-7642-40d8-a65e-f68e4eb0062e">

Reply:
In theory [`str::contains`](https://doc.rust-lang.org/std/primitive.str.html#method.contains) should have the same performance as `starts_with` when the match is at the beginning of the string, right?

Reply:
Switched to contains in [`bcd1250` (#119)](https://github.com/georust/wkt/pull/119/commits/bcd1250b732953cf6edf3d7b8203a0573f662a66)